### ISSUE TITLE
openPMD output fails on AMD MI100

### ISSUE THREAD

**Issue by steindev** [ComputationalRadiationPhysics/picongpu#3624]
I try to get PIConGPU running on [spock](https://docs.olcf.ornl.gov/systems/spock_quick_start_guide.html) cluster at OLCF

For now I had to compile blosc (1.21.0), ADIOS2 (2.7.1) and openPMD (0.13.4) myself.
When running the *LaserWakefield* example I receive the following error during write out of the first simulation data
```
Unhandled exception of type 'St13runtime_error' with message 'expression ((uint64_cu) globalParticleOffset == rp.globalNumParticles) failed in file (/ccs/home/ksteinig/src/picongpu/include/pmacc/../picongpu/plugins/openPMD/WriteSpecies.hpp:173) : ', terminating
```

The exact same setup runs on summit. Do you have an idea what to do?

**cc'ing** @sbastrakov, @psychocoderHPC, @franzpoeschel 

PS: I did not compile `zlib`, as it seems not required. (I would have expected a compile error if it would be required). However the manual lists it as required, so I am not 100% sure.

**EDIT**: the full error log is attach now.

[stderr.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/6538301/stderr.txt)

**Comment by steindev** [comment]
Just to be sure, I compiled `zlib` and ran with it. Error persists. We should probably update the manual, but that is another issue.

**Comment by sbastrakov** [comment]
Are you checking the exact same commit on the two systems @steindev ? It is theoretically possible some recent commit unrelated to the output is causing heap corruption.

**Comment by steindev** [comment]
Yes. summit and spock share the same directories (on all levels). After login to summit I source the summit profile,  `cd` into my setup folder, delete the build folder, recompile and submit via `tbg`.

**Comment by franzpoeschel** [comment]
I think I'll have time to start testing on Spock beginning next week. (Do I need a special access token?)

The error looks like [this `AllGather`](https://github.com/ComputationalRadiationPhysics/picongpu/blob/e17213d472bb259e21aecc001344d54c6b74a9bf/include/picongpu/plugins/openPMD/WriteSpecies.hpp#L398) and [`ConcatListOfFrames`](https://github.com/ComputationalRadiationPhysics/picongpu/blob/e17213d472bb259e21aecc001344d54c6b74a9bf/include/pmacc/particles/operations/ConcatListOfFrames.hpp#L77) are disagreeing on the number of particles. 

Under the assumption that one of both still computes the correct result, we should figure out which one doesn't. Suggestions:
* print both numbers to stdout before the failing line, maybe one of them looks weird
* `ConcatListOfFrames` is arguably the more complex operation, suggesting that it's likelier to be the failing one. By specifying `--dataPreparationStrategy mappedMemory`, you can avoid using it. If that workaround succeeds, we have the culprit.

**Comment by steindev** [comment]
With `--dataPreparationStrategy mappedMemory` the error persists. So it may be the other gather method.

**Comment by franzpoeschel** [comment]
You still get an error or you still get the same error? Normally, the same error shouldn't even be possible since that code path isn't taken.

**Comment by steindev** [comment]
> You still get an error or you still get the same error? Normally, the same error shouldn't even be possible since that code path isn't taken.

Ahh indeed. The line in `WriteSpecies.hpp` changed from 173 to 232 (with mapped memory strategy). Sorry for not looking properly. See 
[stderr_mapped-memory.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/6538656/stderr_mapped-memory.txt)

**Comment by psychocoderHPC** [comment]
> I think I'll have time to start testing on Spock beginning next week. (Do I need a special access token?)

Same access token as you use to access summit.

**Comment by franzpoeschel** [comment]
> > I think I'll have time to start testing on Spock beginning next week. (Do I need a special access token?)
> 
> Same access token as you use to access summit.

Thanks!



> > You still get an error or you still get the same error? Normally, the same error shouldn't even be possible since that code path isn't taken.
> 
> Ahh indeed. The line in `WriteSpecies.hpp` changed from 173 to 232 (with mapped memory strategy). Sorry for not looking properly (but the message is the same :-). See
> [stderr_mapped-memory.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/6538656/stderr_mapped-memory.txt)

Ok, that looks more sensible. I still don't think the error is with the `MPI_Allgather` since that one is basically MPI 101. I thank that there is something more fundamentally wrong with the data loading from GPU, but I think René or Sergei will know more on that than I do.

**Comment by psychocoderHPC** [comment]
I am not sure if https://github.com/ComputationalRadiationPhysics/picongpu/blob/e17213d472bb259e21aecc001344d54c6b74a9bf/include/picongpu/plugins/openPMD/WriteSpecies.hpp#L173 is correct, IMO it compares the global number of particles with the local number of particles. `globalParticleOffset` is a function local variable and initialized with zero and `p.globalNumParticles` is initialized with the global number of particles (from all GPUs) https://github.com/ComputationalRadiationPhysics/picongpu/blob/e17213d472bb259e21aecc001344d54c6b74a9bf/include/picongpu/plugins/openPMD/WriteSpecies.hpp#L422-L430

**Comment by steindev** [comment]
> The assert for the ADIOS strategy should check the locla number of particles as we do in the other strategy.
> 
> https://github.com/ComputationalRadiationPhysics/picongpu/blob/e17213d472bb259e21aecc001344d54c6b74a9bf/include/picongpu/plugins/openPMD/WriteSpecies.hpp#L232

But this is an additional fact, isn't it? I tried the HDF5 strategy (== mappedMemory) and it failed.
And it does not explain, why I/O worked on CUDA accelerators, right?

**Comment by psychocoderHPC** [comment]
> > The assert for the ADIOS strategy should check the locla number of particles as we do in the other strategy.
> > https://github.com/ComputationalRadiationPhysics/picongpu/blob/e17213d472bb259e21aecc001344d54c6b74a9bf/include/picongpu/plugins/openPMD/WriteSpecies.hpp#L232
> 
> But this is an additional fact, isn't it? I tried the HDF5 strategy (== mappedMemory) and it failed.
> And it does not explain, why I/O worked on CUDA accelerators, right?

let me check if mapped memory is working correctly on AMD. I think I tested the last time the adios strategy only.

**Comment by psychocoderHPC** [comment]
I checked the adios IO plugin an https://github.com/ComputationalRadiationPhysics/picongpu/issues/3624#L232 must compare with the local number of particles not with the global number of particles. 
Ahh I think I know why nobody before got this error, @steindev is compiling in debug mode and `PMACC_ASSERT` is disabled in release mode, therefore this error was not visible before, maybe we should change the check to `PMACC_VERIFY` and to this validation check always.


The error with mapped memory needs some more testing. We need to check mapped memory on AMD first. If I remember correctly I had some issues with mapped memory on AMD in the past too.

**Comment by psychocoderHPC** [comment]
I assigned this issue to myself and will fix it for the default strategy.

**Comment by psychocoderHPC** [comment]
Additionally to the bug in IO we figured out that `-DNDEBUG` is not propagated to `hipcc`. This will be fixed in alpaka.

**Comment by psychocoderHPC** [comment]
Issue should be solved with #3626. Both mapping strategies running in my tests on an AMD system without any crash. I have not validated the data.

**Comment by psychocoderHPC** [comment]
The Cmake HIP issue that `-DNDEBUG` is not propagated to hipcc will be solved in alpaak https://github.com/alpaka-group/alpaka/pull/1324 and backported to alpaka 0.6.1 https://github.com/alpaka-group/alpaka/pull/1320 too.

**Comment by steindev** [comment]
I can confirm. The fix in #3626 solves the issue.
